modules/fileout_netcdf/FONcArrayStructureField.cc

```diff
@@ -0,0 +1,366 @@
+            break;
+        }
+
+        case dods_uint16_c: {
+            auto uint16_var = dynamic_cast<UInt16 *>(b);
+            uint16_t uint16_value = uint16_var->value();
+            memcpy(data_buf_ptr,(void*)&uint16_value,d_array_type_size);
+            break;
+        }
+        case dods_int16_c: {
+            auto int16_var = dynamic_cast<Int16 *>(b);
+            int16_t int16_value = int16_var->value();
+            memcpy(data_buf_ptr, (void *) &int16_value, d_array_type_size);
+            break;
+        }
+        case dods_uint32_c: {
+            auto uint32_var = dynamic_cast<UInt32 *>(b);
+            uint32_t uint32_value = uint32_var->value();
+            memcpy(data_buf_ptr,(void*)&uint32_value,d_array_type_size);
+            break;
+        }
+        case dods_int32_c: {
+            auto int32_var = dynamic_cast<Int32 *>(b);
+            int32_t int32_value = int32_var->value();
+            memcpy(data_buf_ptr, (void *) &int32_value, d_array_type_size);
+            break;
+        }
+        case dods_uint64_c: {
+            auto uint64_var = dynamic_cast<UInt64 *>(b);
+            uint64_t uint64_value = uint64_var->value();
+            memcpy(data_buf_ptr,(void*)&uint64_value,d_array_type_size);
+            break;
+        }
+        case dods_int64_c: {
+            auto int64_var = dynamic_cast<Int64 *>(b);
+            int64_t int64_value = int64_var->value();
+            memcpy(data_buf_ptr,(void*)&int64_value,d_array_type_size);
+            break;
+        }
+        case dods_float32_c: {
+            auto float32_var = dynamic_cast<Float32 *>(b);
+            float float32_value = float32_var->value();
+            memcpy(data_buf_ptr,(void*)&float32_value,d_array_type_size);
+            break;
+        }
+        case dods_float64_c: {
+            auto float64_var = dynamic_cast<Float64 *>(b);
+            double float64_value = float64_var->value();
+            memcpy(data_buf_ptr,(void*)&float64_value,d_array_type_size);
+            break;
+        }
+        default:
+            auto err = (string) "file out netcdf structure array: Only support int/float types";
+            throw BESInternalError(err, __FILE__, __LINE__);
+    }
+}
+
+/** @brief returns the name of the array structure field
+ *
+ * @returns The name of the array structure field
+ */
+string
+FONcArrayStructureField::name()
+{
+    return var_name ;
+}
+
+/** @brief returns the netcdf type of the DAP object
+ *
+ * @returns The nc_type of this field.
+ */
+nc_type
+FONcArrayStructureField::type()
+{
+    return d_array_type;
+}
+
+// The following method is adapted from FONcArray.cc.
+FONcDim *
+FONcArrayStructureField::find_sdim(const string &name, int64_t size) {
+
+    FONcDim *ret_dim = nullptr;
+    vector<FONcDim *>::iterator i = FONcArrayStructureField::SDimensions.begin();
+    vector<FONcDim *>::iterator e = FONcArrayStructureField::SDimensions.end();
+    for (; i != e && !ret_dim; i++) {
```

Comment:
Range-baed for.

Also, I don't think you need to test for !empty().

Reply:
This cannot be done with range-based and the !empty() is a key check to ensure the generation of dimension names .